contrib/babelfishpg_common/src/datetime.c

```diff
@@ -663,6 +666,454 @@ datetime_pl_datetime(PG_FUNCTION_ARGS)
+					 * 4. len = 8 : The field is considered as format "YYYY-MM-DD"
+					 * 		and ftype will be modified to "DTK_DATE"
+					 * 5. len = 3,5,7,>8 : Not an valid and should throw error
+					 */
+					switch (len)
+					{
+						case 1:
+						case 2:
+							/*
+							 * Throw an error if next field doesn't exist or
+							 * the next field is not "[ap]m"
+							 */
+							if (i == nf - 1 || ftype[i+1] != DTK_STRING)
+								TIME_IN_ERROR();
+							if (pg_strcasecmp(field[i+1], "am") == 0 || pg_strcasecmp(field[i+1], "pm") == 0)
+							{
+								/*
+								 * For example if the input is "1 am"
+								 * will be converted to "1:00:00 am".
+								 */
+								res = makeStringInfo();
+								appendStringInfo(res, "%s%s", field[i], ":00:00");
+								field[i] = res->data;
+								ftype[i] = DTK_TIME;
+								pfree(res);
+							}
+							else
+								TIME_IN_ERROR();
+							break;
+
+						case 4:
+							/*
+							 * For example if input is "2000" will be converted
+							 * to "2000-01-01".
+							 */
+							res = makeStringInfo();
+							appendStringInfo(res, "%s%s", field[i], "-01-01");
+							field[i] = res->data;
+							ftype[i] = DTK_DATE;
+							pfree(res);
+							break;
+
+						case 6:
+						case 8:
+							/*
+							 * For example if input is "201213" will be converted
+							 * to "20-12-13", and input "20001118" will be
+							 * converted to "2000-11-18"
+							 */
+							res = makeStringInfo();
+							for (int k = 0; k < len; k++)
+							{
+								appendStringInfo(res, "%c", temp_field[k]);
+								if ((len == 6 && ((k + 1) % 2 == 0 && k < 5)) ||
+										(len == 8 && (k == 3 || k == 5)))
+									appendStringInfo(res, "%c", '-');
+							}
+							field[i] = res->data;
+							ftype[i] = DTK_DATE;
+							pfree(res);
+							break;
+
+						default:
+							TIME_IN_ERROR();
+					}
+					break;
+
+				case DTK_DATE:
+					/*
+					 * If the modified_str is of format "<DATE>T<TIME>", then the
+					 * <DATE> should be of format "YYYY-MM-DD". If there are any
+					 * other date formats then an error will be thrown.
+					 */
+					if (i == 0 && nf > 1 && field[1] && pg_strcasecmp(field[1], "t") == 0)
+					{
+						pattern = "^[1-9][0-9]{3}-[0-9]?[0-9]-[0-9]?[0-9]$";
+						if (regcomp(&time_regex, pattern, REG_EXTENDED) != 0)
+							ereport(ERROR,
+										(errcode(ERRCODE_INTERNAL_ERROR),
+										errmsg("time format internal error")));
+						if (regexec(&time_regex, field[i], 0, NULL, 0) != 0)
+							TIME_IN_ERROR();
+						regfree(&time_regex);
+					}
+					/*
+					 * If the date of modified_str is of any formats {"Mon{/.-}yyyy{/.-}dd",
+					 * "Mon{/.-}dd{/.-}yyyy", "DD{/.-}MM{/.-}YYYY", "DD{/.-}YYYY{/.-}MM"}
+					 * then these dates shouldn't be supported and an error should be thrown.
+					 * Supported date formats are {"DD-MON-YYYY", "YYYY-MON-DD",
+					 * "MM{-/.}DD{-/.}YYYY", "YYYY{-/.}MM{-/.}DD"}.
+					 */
+					else
+					{
+						for (int k = 0 ; k < 3; k++)
```

Comment:
The field can be of pattern like  {"Mon{/.-}yyyy{/.-}dd", "Mon{/.-}dd{/.-}yyyy", "DD{/.-}MM{/.-}YYYY", "DD{/.-}YYYY{/.-}MM", "DD-MON-YYYY", "YYYY-MON-DD", "MM{-/.}DD{-/.}YYYY", "YYYY{-/.}MM{-/.}DD"}. It will be hard to come to conclusion of what can be the first part of field before any of [/.-]. That's why this for loop will get the part before first occurance of [/.-]. Which will be helpful in furthur analysis.